torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -1031,6 +1030,335 @@ std::pair<TensorDomain*, TensorDomain*> TensorDomain::rFactor(
+  //! \param e element e to find the equiv class for
+  //! \returns the equivalent class that e belongs to
+  //!
+  int fixedPoint(T e) const {
+    // Handles case when i doesn't have an equivalence class
+    TORCH_INTERNAL_ASSERT(entry_map.count(e));
+
+    // Use fixed point as a representation for the equiv class
+    return fixedPoint(entry_map.at(e));
+  }
+
+  //! Utility to create a new equiv class for i
+  //
+  //! \param i Element i to create the equiv class for
+  void createPoint(T i) {
+    entry_map[i] = next_index_;
+    set_map.push_back(next_index_++);
+    weights.push_back(1);
+  }
+
+ private:
+  // Internal representation of the equivalence class as integers
+  // set_map implements the "parent" relationship
+  std::vector<int> set_map;
+  // weights is used for preliminary perf optimization
+  std::vector<int> weights;
+
+  // Map the input of type T to its equivalence class
+  std::unordered_map<T, int> entry_map;
+
+  // Running counter for generating new index when
+  // creating new equiv classes
+  int next_index_ = 0;
+};
+
+//! Concretize broadcast axes, i.e. identifying a non-broadcast
+//! IterDomain that the broadcast IterDomain can map to.
+//!
+//! This traversal processes root domains only, concretization works by
+//! inspecting pointwise ops, e.g. : T2 [i0,i1] = T1[i0,B0] + T0[i0,i1]
+//! will concretize axis B0 to i1
+//!
+class ConcretizeDomain : private BackwardVisitor {
+ public:
+  //! Traverses the graph backward from outputs
+  //! to identify all concretizing opportunities
+  //!
+  explicit ConcretizeDomain(Fusion* fusion) {
+    traverseFrom(fusion, fusion->outputs(), false);
+  }
+
+  //! API call to run the concretize pass and return the
+  //! axis that bcast_dom concretizes to
+  //!
+  static const IterDomain* getConcreteDomain(IterDomain* bcast_dom) {
+    ConcretizeDomain cd(bcast_dom->fusion());
+
+    // Remove this assertion once we support broadcast on output
+    TORCH_INTERNAL_ASSERT(cd.canConcretize(bcast_dom));
+    return cd.concretized(bcast_dom);
+  }
+
+  // Returns true if either id is not a broadcast or
+  // the traversal has found a concretized axis for id
+  bool canConcretize(IterDomain* id) const {
+    return !id->isBroadcast() || bcast_domain_map_.count(id);
+  }
+
+  // Returns the concretized id recorded from traversal
+  IterDomain* concretized(IterDomain* id) const {
+    TORCH_INTERNAL_ASSERT(canConcretize(id));
+    if (!id->isBroadcast())
+      return id;
+    return bcast_domain_map_.at(id);
+  }
+
+ private:
+  // Utility to inspect a pointwise operator and
+  // record concretize opportunities
+  void concretizePwOp(Expr* e);
+
+  // Utility to record new concretize opportunity
+  void concretizeTo(IterDomain* id, IterDomain* To) {
+    TORCH_INTERNAL_ASSERT(id->isBroadcast() && !To->isBroadcast());
+    bcast_domain_map_[id] = concretized(To);
+  }
+
+  void handle(ReductionOp* rop) override {
+    concretizePwOp(rop->asExpr());
```

Comment:
`->asExpr()` is not needed in any of the handle() functions (since the conversion to the Expr base class is implicit)

Reply:
Updated, thanks!